File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def check_diff_one(word1, word2):
    diff_cnt = 0
    
    for (c1, c2) in zip(word1, word2):
        if c1 != c2:
            diff_cnt += 1
            
        if diff_cnt > 1:
            return False
            
    return True if diff_cnt == 1 else False


def solution(begin, target, words):
    if target not in words:
        return 0
    
    answer = float('inf')
    n = len(words)
    
    visited = [False] * n
    
    def dfs(visited, curr_word, cnt):
        nonlocal answer
        
        if curr_word == target:
            return cnt
        
        for i, word in enumerate(words):
            if check_diff_one(curr_word, word) and not visited[i]:
                visited[i] = True
                
                answer = min(answer, dfs(visited, word, cnt + 1))
                
                visited[i] = False
                
        return answer
        
    return dfs(visited, begin, 0)

File: 프로그래머스/3/43163. 단어 변환/단어 변환.kt
class Solution {
    data class Point(val word: String, val step: Int)
    
    fun isDiffOne(word1: String, word2: String): Boolean {
        var cnt = 0
        
        for ((a, b) in word1.zip(word2)) {
            if (a == b) continue
            
            cnt++
            
            if (cnt > 1) return false
        }
        
        return true
    }
    
    fun solution(begin: String, target: String, words: Array<String>): Int {
        if (target !in words) return 0

        val queue = ArrayDeque<Point>()
        val visited = mutableSetOf<String>()
        
        queue.addLast(Point(begin, 0))
        visited.add(begin)
        
        while (queue.isNotEmpty()) {
            val (word, step) = queue.removeFirst()
            
            if (word == target) return step
            
            for (diffWord in words) {
                if (diffWord !in visited && isDiffOne(word, diffWord)) {
                    queue.addLast(Point(diffWord, step + 1))
                    visited.add(diffWord)
                }
            }
        }
        
        error("종료")
    }
}